File: packages/meal-plan-engine/src/slot-allocator.ts
import type { MacroTargets, MealSlot, SlotAllocation } from "@caloriehero/shared-types";

const PERCENTAGE_SUM_TOLERANCE = 0.01;

/**
 * Distributes daily macro targets across meal slots proportionally.
 *
 * Validates that percentages sum to approximately 1.0 (within 0.01 tolerance).
 * Applies rounding correction to ensure the allocated macros precisely sum
 * to the daily targets.
 *
 * @throws Error if percentages do not sum to ~1.0
 */
export function allocateSlots(
  dailyTargets: MacroTargets,
  slots: { slot: MealSlot; percentage: number }[]
): SlotAllocation[] {
  if (slots.length === 0) {
    throw new Error("At least one slot must be provided");
  }

  const totalPercentage = slots.reduce((sum, s) => sum + s.percentage, 0);
  if (Math.abs(totalPercentage - 1.0) > PERCENTAGE_SUM_TOLERANCE) {
    throw new Error(
      `Slot percentages must sum to 1.0 (got ${totalPercentage.toFixed(4)})`
    );
  }

  // Calculate raw allocations (may have floating point imprecision)
  const rawAllocations = slots.map((s) => ({
    slot: s.slot,
    percentage: s.percentage,
    targets: {
      calories: dailyTargets.calories * s.percentage,
      protein: dailyTargets.protein * s.percentage,
      carbs: dailyTargets.carbs * s.percentage,
      fat: dailyTargets.fat * s.percentage,
    },
  }));

  // Apply rounding correction: round each allocation and fix the last slot
  // so that sum exactly equals daily targets.
  const rounded = rawAllocations.map((a) => ({
    slot: a.slot,
    percentage: a.percentage,
    targets: {
      calories: Math.round(a.targets.calories * 100) / 100,
      protein: Math.round(a.targets.protein * 100) / 100,
      carbs: Math.round(a.targets.carbs * 100) / 100,
      fat: Math.round(a.targets.fat * 100) / 100,
    },
  }));

  // Fix rounding errors by adjusting the last slot
  if (rounded.length > 0) {
    const sumCalories = rounded.slice(0, -1).reduce((s, a) => s + a.targets.calories, 0);
    const sumProtein = rounded.slice(0, -1).reduce((s, a) => s + a.targets.protein, 0);
    const sumCarbs = rounded.slice(0, -1).reduce((s, a) => s + a.targets.carbs, 0);
    const sumFat = rounded.slice(0, -1).reduce((s, a) => s + a.targets.fat, 0);

    const last = rounded[rounded.length - 1];
    if (last !== undefined) {
      last.targets = {
        calories: Math.round((dailyTargets.calories - sumCalories) * 100) / 100,
        protein: Math.round((dailyTargets.protein - sumProtein) * 100) / 100,
        carbs: Math.round((dailyTargets.carbs - sumCarbs) * 100) / 100,
        fat: Math.round((dailyTargets.fat - sumFat) * 100) / 100,
      };
    }
  }

  return rounded as SlotAllocation[];
}

File: backend/app/engine/slot_allocator.py
"""Distribute daily macro targets across meal slots."""

from app.engine.types import MacroTargets, SlotAllocation

PERCENTAGE_SUM_TOLERANCE = 0.01


def allocate_slots(
    daily_targets: MacroTargets,
    slots: list[dict[str, object]],
) -> list[SlotAllocation]:
    """Distribute daily macro targets across meal slots proportionally.

    Validates that percentages sum to ~1.0 (within 0.01 tolerance).
    Applies rounding correction to ensure allocated macros sum exactly to daily targets.

    Raises:
        ValueError: If no slots provided or percentages don't sum to ~1.0.
    """
    if not slots:
        raise ValueError("At least one slot must be provided")

    total_percentage = sum(float(s["percentage"]) for s in slots)  # type: ignore[arg-type]
    if abs(total_percentage - 1.0) > PERCENTAGE_SUM_TOLERANCE:
        raise ValueError(
            f"Slot percentages must sum to 1.0 (got {total_percentage:.4f})"
        )

    # Calculate raw allocations with rounding
    rounded: list[SlotAllocation] = []
    for s in slots:
        slot = str(s["slot"])
        pct = float(s["percentage"])  # type: ignore[arg-type]
        targets = MacroTargets(
            calories=round(daily_targets.calories * pct, 2),
            protein=round(daily_targets.protein * pct, 2),
            carbs=round(daily_targets.carbs * pct, 2),
            fat=round(daily_targets.fat * pct, 2),
        )
        rounded.append(SlotAllocation(slot=slot, percentage=pct, targets=targets))  # type: ignore[arg-type]

    # Fix rounding errors by adjusting the last slot
    if len(rounded) > 1:
        sum_calories = sum(a.targets.calories for a in rounded[:-1])
        sum_protein = sum(a.targets.protein for a in rounded[:-1])
        sum_carbs = sum(a.targets.carbs for a in rounded[:-1])
        sum_fat = sum(a.targets.fat for a in rounded[:-1])

        last = rounded[-1]
        corrected_targets = MacroTargets(
            calories=round(daily_targets.calories - sum_calories, 2),
            protein=round(daily_targets.protein - sum_protein, 2),
            carbs=round(daily_targets.carbs - sum_carbs, 2),
            fat=round(daily_targets.fat - sum_fat, 2),
        )
        rounded[-1] = SlotAllocation(
            slot=last.slot, percentage=last.percentage, targets=corrected_targets
        )

    return rounded
